src/core/frames/frame_controller.js

```diff
@@ -354,10 +354,10 @@ export class FrameController {
       const pageSnapshot = PageSnapshot.fromElement(frame).clone()
       const { visitCachedSnapshot } = frame.delegate
 
-      frame.delegate.fetchResponseLoaded = (fetchResponse) => {
+      frame.delegate.fetchResponseLoaded = async (fetchResponse) => {
         if (frame.src) {
           const { statusCode, redirected } = fetchResponse
-          const responseHTML = frame.ownerDocument.documentElement.outerHTML
+          const responseHTML = await fetchResponse.responseHTML
```

Comment:
This will likely cause a full page reload because the `responseHTML` will only have a minimal `<head>` and therefore the tracked elements won't be identical.

Reply:
Thank you for catching this @domchristie. I'm not sure how to best resolve the issue mentioned in #1047. Do you have an idea on what to change?

Reply:
@seanpdoyle Can this line be reverted or will that break the intention of https://github.com/hotwired/turbo/pull/867 / https://github.com/hotwired/turbo-rails/pull/428?

Otherwise, I'm beginning to feel that the different visit usages warrant a different modelling.

> For example:
>
> 1. A standard Visit that navigates from one location to the next
> 2. A same-page anchor Visit
> 3. A redirected Visit
> 4. A Frame Visit (promoted to a Visit using data-turbo-action
>
> Each has different expectations and outcomes, which hints that they could be different classes that share the same API
> https://github.com/hotwired/turbo/pull/1044#issuecomment-1780692991

This would be a fairly big refactor, but might be worth it?

Reply:
https://github.com/hotwired/turbo/pull/430 explores a `FrameVisit`, but I'm unsure of what changes would be necessary to incorporate a fix for this issue.

> Can this line be reverted or will that break the intention of #867 / [hotwired/turbo-rails#4](https://github.com/hotwired/turbo-rails/pull/428)

I'm unsure of the implications of reverting it. Is there a conditional that could be introduced to support responses without a `<head>` **and** responses from newer `turbo-rails` versions that send a minimal head?

Reply:
> I'm unsure of the implications of reverting it.

It seems like v7.3.0 includes https://github.com/hotwired/turbo/pull/867 so reverting _should_ work (although I'm not sure how!?) I'm not 

Off the top of my head (pun intended ;), I'm not sure how `HeadSnapshot` could distinguish between a "minimal layout" head from a frame visit and standard visit, where the page's tracked assets have actually changed

Reply:
> but I'm unsure of what changes would be necessary to incorporate a fix for this issue.

It doesn't feel right to me that we call `session.proposeVisitIfNavigatedWithAction` after rendering a frame response. (In the same way it doesn't feel right to propose a second visit after rendering a redirected response.) In these cases, we're not "visiting" a new location, we're just using the side-effects of the visit lifecycle.

In terms of what I had in mind for implementing different visit types: I was thinking that each Visit type could share an interface, and each implementation could respond accordingly (e.g. a `RedirectVisit` wouldn't dispatch any events because the Visit lifecycle events have already been broadcast by the initial Visit.)

Reply:
Any progress?

Reply:
This change broke Restoration Visits with `turbo-frame data-turbo-action="advance"`.

see: 
- [Issue: Turbo frame history not navigating properly w/ back button after upgrading from 7 to 8](https://github.com/hotwired/turbo/issues/1241)
- [PR: Support back button (restore) even if page loaded in frame does not contain matching HEAD](https://github.com/hotwired/turbo/pull/1281#issuecomment-2185427909)

Is this something we can revert or is there a different approach we can take?

Reply:
also see #1300